PWGEM/PhotonMeson/Tasks/CheckMCV0.cxx

```diff
@@ -10,275 +10,533 @@
       return false;
     }
     if (abs(track.eta()) > maxeta) {
+      registry.fill(HIST("CheckV0Leg"), checkV0legEnum::MAXETA);
       return false;
     }
-
-    if (abs(track.dcaXY()) < dcamin || dcamax < abs(track.dcaXY())) {
+    if (abs(track.dcaXY()) < dcamin) {
+      registry.fill(HIST("CheckV0Leg"), checkV0legEnum::DCAMIN);
       return false;
     }
-    if (!track.hasITS() && !track.hasTPC()) {
+    if (abs(track.dcaXY()) > dcamax) {
+      registry.fill(HIST("CheckV0Leg"), checkV0legEnum::DCAMAX);
       return false;
     }
-    if (track.tpcNClsCrossedRows() < minTPCNCls) {
+    if (track.tpcChi2NCl() > maxChi2TPC) {
+      registry.fill(HIST("CheckV0Leg"), checkV0legEnum::TPCCHI2NCL);
+      return false;
+    }
+    if (track.tpcNClsCrossedRows() < minCrossedRowsTPC) {
+      registry.fill(HIST("CheckV0Leg"), checkV0legEnum::MINCROSSEDROWSTPC);
       return false;
     }
     return true;
   }
 
   template <typename TTrack>
-  bool isITSTPCMatchedTrack(TTrack const& track)
+  inline void checkPassed(TTrack const& track0, TTrack const& track1)
   {
-    return track.hasITS() && track.hasTPC();
-  }
-
-  template <typename TTrack>
-  bool isTPConlyTrack(TTrack const& track)
-  {
-    return !track.hasITS() && track.hasTPC();
-  }
-
-  template <typename TTrack>
-  bool isITSonlyTrack(TTrack const& track)
-  {
-    return track.hasITS() && !track.hasTPC();
-  }
-
-  template <typename TTrack>
-  bool isITSTPC_ITSTPC(TTrack const& track0, TTrack const& track1)
-  {
-    return isITSTPCMatchedTrack(track0) && isITSTPCMatchedTrack(track1);
-  }
-
-  template <typename TTrack>
-  bool isITSTPC_TPConly(TTrack const& track0, TTrack const& track1)
-  {
-    return (isITSTPCMatchedTrack(track0) && isTPConlyTrack(track1)) || (isITSTPCMatchedTrack(track1) && isTPConlyTrack(track0));
+    v0TypesPassed.fill(false); // reset
+    v0TypesPassed[0] = isITSTPC_ITSTPC(track0, track1);
+    v0TypesPassed[1] = isTPConly_TPConly(track0, track1);
+    v0TypesPassed[2] = isITSonly_ITSonly(track0, track1);
+    v0TypesPassed[3] = isITSTPC_TPConly(track0, track1);
+    v0TypesPassed[4] = isITSTPC_ITSonly(track0, track1);
+    v0TypesPassed[5] = isTPConly_ITSonly(track0, track1);
+    for (int i = 0; i < v0TypesPassed.size(); ++i) {
+      if (v0TypesPassed[i]) {
+        registry.fill(HIST("V0TypeCounter"), i + 1);
+        treeData.type = v0TypesEnum(i);
+      }
+    }
   }
 
-  template <typename TTrack>
-  bool isITSTPC_ITSonly(TTrack const& track0, TTrack const& track1)
+  void setTree()
   {
-    return (isITSTPCMatchedTrack(track0) && isITSonlyTrack(track1)) || (isITSTPCMatchedTrack(track1) && isITSonlyTrack(track0));
-  }
 
-  template <typename TTrack>
-  bool isTPConly_TPConly(TTrack const& track0, TTrack const& track1)
-  {
-    return isTPConlyTrack(track0) && isTPConlyTrack(track1);
+    tree.setObject(new TTree("checkMC", "checkMC"));
+
+    // Reconstructed data
+    tree->Branch("trkPosPt", &treeData.trkPosPt, "trkPosPt/F");
```

Comment:
If you make a tree, can you move this task to TableProducer and call Produce<your skimmed data>?